src/__tests__/starWarsStreamQuery-test.js

```diff
@@ -0,0 +1,200 @@
+          hero {
+            friends @stream(initial_count: 0, label: "HeroFriends") {
+              id
+              name
+            }
+          }
+        }
+      `;
+      const result = await graphql(StarWarsSchema, query);
+      expect(result).to.deep.equal({
+        errors: [
+          {
+            message: 'Unknown directive "@stream".',
+            locations: [{ line: 4, column: 21 }],
+          },
+        ],
+      });
+    });
+    it('Can disable @stream using if argument', async () => {
+      const query = `
+        query HeroFriendsQuery {
+          hero {
+            friends @stream(initial_count: 0, label: "HeroFriends", if: false) {
+              id
+              name
+            }
+          }
+        }
+      `;
+      const result = await graphql(StarWarsSchemaDeferStreamEnabled, query);
+      expect(result).to.deep.equal({
+        data: {
+          hero: {
+            friends: [
+              {
+                id: '1000',
+                name: 'Luke Skywalker',
+              },
+              {
+                id: '1002',
+                name: 'Han Solo',
+              },
+              {
+                id: '1003',
+                name: 'Leia Organa',
+              },
+            ],
+          },
+        },
+      });
+    });
+  });
+
+  describe('Basic Queries', () => {
+    it('Can @stream an array field', async () => {
+      const query = `
+        query HeroFriendsQuery {
+          hero {
+            friends @stream(initial_count: 2, label: "HeroFriends") {
+              id
+              name
+            }
+          }
+        }
+      `;
+      const result = await graphql(StarWarsSchemaDeferStreamEnabled, query);
+      const { patches: patchesIterable, ...initial } = result;
+      expect(initial).to.deep.equal({
+        data: {
+          hero: {
+            friends: [
+              {
+                id: '1000',
+                name: 'Luke Skywalker',
+              },
+              {
+                id: '1002',
+                name: 'Han Solo',
+              },
+            ],
+          },
+        },
+      });
+
+      const patches = [];
+
+      if (patchesIterable) {
+        await forAwaitEach(patchesIterable, patch => {
+          patches.push(patch);
+        });
+      }
+
+      expect(patches).to.have.lengthOf(1);
+      expect(patches[0]).to.deep.equal({
+        label: 'HeroFriends',
+        path: ['hero', 'friends', 2],
+        data: {
+          id: '1003',
+          name: 'Leia Organa',
+        },
+      });
+    });
+    it('Can @stream multiple selections on the same field', async () => {
+      const query = `
+        query HeroFriendsQuery {
+          hero {
+            friends {
+              id
+            }
+            ...FriendsName
+            ...FriendsAppearsIn
+          }
+        }
+        fragment FriendsName on Character {
+          friends @stream(label: "nameLabel", initial_count: 1) {
```

Comment:
At Facebook, we actually invalidate a query that looks like this. In practice, this is more often a mistake developers make rather than an intentional decision.

To allow developers to create multiple streams on the same field, or to both stream and not stream, we take advantage of the "alias" feature in GraphQL and do something like this:

```
friendsWithName: friends @stream(label: "namedLabel", initial_count: 1)
```

Reply:
@Keweiqu I updated the PR to add a validation rule that forbids streaming on the same field unless the label and initial_count are the same. Does this more closely match the implementation at Facebook?

https://github.com/graphql/graphql-js/pull/2319/commits/c29e1e327e796e53047a7d224eac4ea4eb9d48ed#diff-a49ecb51ef73494dfb5db4effdf536a6R601

Reply:
@robrichard Thanks for adding the validation rule. I do think this validation is universally beneficial no matter which infrastructure developers are working with.